pkg/vault/test/foundry/BatchRouterE2E.t.sol

```diff
@@ -373,4 +373,555 @@ contract BatchRouterE2ETest is BaseBatchRouterE2ETest {
+
+        UniversalSwapPath[] memory path = new UniversalSwapPath[](1);
+        path[0] = UniversalSwapPath({
+            tokenIn: IERC20(weth),
+            steps: new SwapPathStep[](3),
+            givenAmount: givenAmount,
+            limit: limit
+        });
+        path[0].steps[0] = SwapPathStep({
+            pool: getPool(weth, dai),
+            tokenOut: IERC20(getPool(weth, dai)),
+            isBuffer: false
+        });
+        path[0].steps[1] = SwapPathStep({ pool: getPool(weth, dai), tokenOut: dai, isBuffer: false });
+        path[0].steps[2] = SwapPathStep({ pool: getPool(dai, usdc), tokenOut: usdc, isBuffer: false });
+
+        generateSimpleDiffs(path, kind);
+
+        uint256 ethAmount = kind == SwapKind.EXACT_IN ? DEFAULT_AMOUNT : limit;
+        testSwap(path, kind, wethIsEth, wethIsEth ? ethAmount : 0, REMOVE_LIQUIDITY_DELTA);
+    }
+
+    // ------ testExitSwapSinglePathAndFinalRemoveLiquidityStep ------
+    function testExitSwapExactInSinglePathAndFinalRemoveLiquidityStep() public {
+        _testExitSwapSinglePathAndFinalRemoveLiquidityStep(SwapKind.EXACT_IN, false);
+    }
+
+    function testExitSwapExactInSinglePathAndFinalRemoveLiquidityStepETH() public {
+        _testExitSwapSinglePathAndFinalRemoveLiquidityStep(SwapKind.EXACT_IN, true);
+    }
+
+    function testExitSwapExactOutSinglePathAndFinalRemoveLiquidityStep() public {
+        _testExitSwapSinglePathAndFinalRemoveLiquidityStep(SwapKind.EXACT_OUT, false);
+    }
+
+    function testExitSwapExactOutSinglePathAndFinalRemoveLiquidityStepETH() public {
+        _testExitSwapSinglePathAndFinalRemoveLiquidityStep(SwapKind.EXACT_OUT, true);
+    }
+
+    function _testExitSwapSinglePathAndFinalRemoveLiquidityStep(SwapKind kind, bool wethIsEth) internal virtual {
+        uint256 limit = kind == SwapKind.EXACT_IN ? DEFAULT_AMOUNT - REMOVE_LIQUIDITY_ROUNDING_ERROR : DEFAULT_AMOUNT;
+        uint256 givenAmount = kind == SwapKind.EXACT_IN
+            ? DEFAULT_AMOUNT
+            : DEFAULT_AMOUNT - REMOVE_LIQUIDITY_ROUNDING_ERROR;
+
+        UniversalSwapPath[] memory path = new UniversalSwapPath[](1);
+        path[0] = UniversalSwapPath({
+            tokenIn: IERC20(dai),
+            steps: new SwapPathStep[](2),
+            givenAmount: givenAmount,
+            limit: limit
+        });
+        path[0].steps[0] = SwapPathStep({
+            pool: getPool(weth, dai),
+            tokenOut: IERC20(getPool(weth, dai)),
+            isBuffer: false
+        });
+        path[0].steps[1] = SwapPathStep({ pool: getPool(weth, dai), tokenOut: weth, isBuffer: false });
+
+        generateSimpleDiffs(path, kind);
+        testSwap(path, kind, wethIsEth, 0, REMOVE_LIQUIDITY_DELTA);
+    }
+
+    // ------ testExitSwapMultiPathAndFinalRemoveLiquidityStep ------
+    function testExitSwapExactInMultiPathAndFinalRemoveLiquidityStep() public {
+        _testExitSwapMultiPathAndFinalRemoveLiquidityStep(SwapKind.EXACT_IN, false);
```

Comment:
Consider defining constants like: USE_ETH_TRUE, USE_ETH_FALSE, and use them instead of true/false, to self-document what the flags are for.